rust/worker/src/index/metadata/types.rs

```diff
@@ -0,0 +1,281 @@
+    NotInTransaction,
+}
+
+impl ChromaError for MetadataIndexError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            MetadataIndexError::NotFoundError => ErrorCodes::InvalidArgument,
+            MetadataIndexError::InTransaction => ErrorCodes::InvalidArgument,
+            MetadataIndexError::NotInTransaction => ErrorCodes::InvalidArgument,
+        }
+    }
+}
+
+pub(crate) enum MetadataIndexValue {
+    String(String),
+    Float(f32),
+    Bool(bool),
+}
+
+pub(crate) trait MetadataIndex {
+    fn begin_transaction(&mut self) -> Result<(), Box<dyn ChromaError>>;
+    fn commit_transaction(&mut self) -> Result<(), Box<dyn ChromaError>>;
+
+    // Must be in a transaction to put or delete.
+    fn set(&mut self, key: &str, value: MetadataIndexValue, offset_id: usize) -> Result<(), Box<dyn ChromaError>>;
+    // Can delete anything -- if it's not in committed state the delete will be silently discarded.
+    fn delete(&mut self, key: &str, value: MetadataIndexValue, offset_id: usize) -> Result<(), Box<dyn ChromaError>>;
+
+    // Always reads from committed state.
+    fn get(&self, key: &str, value: MetadataIndexValue) -> Result<RoaringBitmap, Box<dyn ChromaError>>;
+}
+
+struct BlockfileMetadataIndex {
+    blockfile: Box<dyn Blockfile>,
+    in_transaction: bool,
+    uncommitted_rbms: HashMap<BlockfileKey, RoaringBitmap>,
+}
+
+impl BlockfileMetadataIndex {
+    pub fn new() -> Self {
+        BlockfileMetadataIndex {
+            blockfile: Box::new(HashMapBlockfile::open(&"in-memory").unwrap()),
+            in_transaction: false,
+            uncommitted_rbms: HashMap::new(),
+        }
+    }
+
+    fn look_up_key_and_populate_uncommitted_rbms(&mut self, key: &BlockfileKey) -> Result<(), Box<dyn ChromaError>> {
+        if !self.uncommitted_rbms.contains_key(&key) {
+            match self.blockfile.get(key.clone()) {
+                Ok(Value::RoaringBitmapValue(rbm)) => {
+                    self.uncommitted_rbms.insert(key.clone(), rbm);
+                },
+                _ => {
+                    let rbm = RoaringBitmap::new();
+                    self.uncommitted_rbms.insert(key.clone(), rbm);
+                },
+            };
+        }
+        Ok(())
+    }
+}
+
+impl MetadataIndex for BlockfileMetadataIndex {
+    fn begin_transaction(&mut self) -> Result<(), Box<dyn ChromaError>> {
+        if self.in_transaction {
+            return Err(Box::new(MetadataIndexError::InTransaction));
+        }
+        self.blockfile.begin_transaction()?;
+        self.in_transaction = true;
+        Ok(())
+    }
+
+    fn commit_transaction(&mut self) -> Result<(), Box<dyn ChromaError>> {
+        if !self.in_transaction {
+            return Err(Box::new(MetadataIndexError::NotInTransaction));
+        }
+        for (key, rbm) in &self.uncommitted_rbms {
+            // TODO we shouldn't clone.
+            self.blockfile.set(key.clone(), Value::RoaringBitmapValue(rbm.clone()));
```

Comment:
you may be able to use https://doc.rust-lang.org/std/collections/hash_map/struct.Drain.html

Reply:
Exactly what I wanted, ty